pkg/kubelet/cm/dra/state/state_checkpoint.go

```diff
@@ -0,0 +1,109 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package state
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/kubelet/checkpointmanager"
+	"k8s.io/kubernetes/pkg/kubelet/checkpointmanager/errors"
+)
+
+var _ CheckpointState = &stateCheckpoint{}
+
+// CheckpointState interface provides to get and store state
+type CheckpointState interface {
+	Get() (ClaimInfoStateList, error)
+	Store(claimInfoStateList ClaimInfoStateList) error
+}
+
+// ClaimInfoState is used to store claim info state in a checkpoint
+type ClaimInfoState struct {
+	// name of the DRA driver
+	DriverName string
+
+	// ClaimUID is an UID of the resource claim
+	ClaimUID types.UID
+
+	// ClaimName is a name of the resource claim
+	ClaimName string
+
+	// Namespace is a claim namespace
+	Namespace string
+
+	// PodUIDs is a set of pod UIDs that reference a resource
+	PodUIDs sets.Set[string]
+
+	// CdiDevices is a list of CDI devices returned by the
+	// GRPC API call NodePrepareResource
+	CdiDevices []string
+}
+
+type stateCheckpoint struct {
+	sync.RWMutex
+	checkpointManager checkpointmanager.CheckpointManager
+	checkpointName    string
+}
+
+// NewCheckpointState creates new State for keeping track of claim info  with checkpoint backend
+func NewCheckpointState(stateDir, checkpointName string) (*stateCheckpoint, error) {
+	checkpointManager, err := checkpointmanager.NewCheckpointManager(stateDir)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize checkpoint manager: %v", err)
+	}
+	stateCheckpoint := &stateCheckpoint{
+		checkpointManager: checkpointManager,
+		checkpointName:    checkpointName,
+	}
+
+	return stateCheckpoint, nil
+}
+
+// get state from a checkpoint and creates it if it doesn't exist
+func (sc *stateCheckpoint) Get() (ClaimInfoStateList, error) {
+	sc.Lock()
+	defer sc.Unlock()
+	var err error
+
+	checkpoint := NewDRAManagerCheckpoint()
+	if err = sc.checkpointManager.GetCheckpoint(sc.checkpointName, checkpoint); err != nil {
+		if err == errors.ErrCheckpointNotFound {
+			sc.Store(ClaimInfoStateList{})
+			return ClaimInfoStateList{}, nil
+		}
+		return nil, err
+	}
+
+	return checkpoint.Entries, nil
+}
```

Comment:
This would be cleaner as:
```suggestion
func (sc *stateCheckpoint) GetOrCreate() (ClaimInfoStateList, error) {
	sc.Lock()
	defer sc.Unlock()

	checkpoint := NewDRAManagerCheckpoint()
	
	err := sc.checkpointManager.GetCheckpoint(sc.checkpointName, checkpoint)
	if err == errors.ErrCheckpointNotFound {
		sc.Store(ClaimInfoStateList{})
		return ClaimInfoStateList{}, nil
	}
	if err != nil {
		return nil, err
	}

	return checkpoint.Entries, nil
}
```

Reply:
done